File: Search2DMatrix.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ConsoleApp1
{
    class Search2DMatrix
    {
        public bool SearchMatrix(int[][] matrix, int target)
        {
            if (matrix == null || matrix.Length == 0)
                return false;

            int m = matrix.Length;
            int n = matrix[0].Length;

            int start = 0;
            int end = m * n - 1;

            while (start <= end)
            {
                int mid = start + (end - start) / 2;
                int row = mid / n;
                int column = mid % n;
                if (matrix[row][column] == target)
                    return true;

                if (matrix[row][column] < target)
                    start = mid + 1;
                else
                    end = mid - 1;
            }

            return false;
        }
    }
}

File: src/Search2DMatrix.java
// Time Complexity :O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Search2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {

        int row = matrix.length;
        int column = matrix[0].length;
        int l = 0;
        int h = row-1;
        while(l<=h)
        {
            int mid = l + (h-l)/2;
            if(matrix[mid][column-1] > target) {
                if(h<row && h>=0)
                    h = mid-1;
            }
            else if(matrix[mid][column-1] < target){
                if(l+1<row && l+1>=0)
                    l = mid+1;
                else{
                    break;
                }
            }
            else{
                return true;
            }
        }
        row = l;
        l =0;
        h =column-1;
        while(l<=h)
        {
            int mid = l + (h-l)/2;
            if(matrix[row][mid] > target)
            {
                h = mid-1;
            }
            else if(matrix[row][mid] < target){
                l = mid+1;
            }
            else{
                return true;
            }
        }
        return false;
    }
}